File: 0953. Verifying an Alien Dictionary/Solution0953.cpp
#include<iostream>
#include<vector>
#include<string>
#include<unordered_map>

using namespace std;

class Solution9553 {
public:
    bool isAlienSorted(vector<string>& words, string order) {
        unordered_map<char, int> ordDict;
        for (int i = 0; i < order.length(); i++) {
            ordDict[order[i]] = i;
        }
        for (int i = 0; i < words.size()-1; i++) {
            if (!check(ordDict, words[i], words[i+1])) {
                return false;
            }
        }
        return true;
    }

    bool check(unordered_map<char, int>& wordDict, string w1, string w2) {
        int limsize = min(w1.length(), w2.length());
        for (int i = 0; i < limsize; i++) {
            if (w1[i] != w2[i]) {
                if (wordDict[w1[i]] > wordDict[w2[i]]) {
                    return false;
                } else {
                    return true;
                }
            } 
        }
        if (w1.length() > w2.length()) return false; //!!!
        return true;
    }
};

File: 0953. Verifying an Alien Dictionary/Solution0953.py
class Solution0953:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        dict_map = {}
        for i in range(len(order)):
            dict_map[order[i]] = i
        for i in range(len(words)-1):
            if not self.check(dict_map, words[i], words[i+1]):
                return False
        return True
    
    def check(self, dict_map, word1, word2) -> bool:
        for i in range(min(len(word1), len(word2))):
            if word1[i] != word2[i]:
                if dict_map[word1[i]] > dict_map[word2[i]]:
                    return False
                else:
                    return True
        if len(word1) > len(word2):
            return False 
        return True